### ISSUE TITLE
Incoming request body fails to be streamed to outgoing client request body

### ISSUE THREAD

**Issue by valeriansaliou** [hyperium/hyper#1291]
Running into issues using Hyper w/ a special use case. I'm building [an OSS REST API caching proxy](https://github.com/valeriansaliou/bloom), which receives HTTP requests to Hyper and needs to proxy them to the "wrapped" API server by emitting a mirrored request on the other end, using Hyper's HTTP client.

All works great w/ GET, HEAD, etc. requests (any request w/ an empty body). However, once the body is filled w/ data (eg. POST, PATCH, PUT, etc.), requests hang forever.

I expected this to work, as I'm just passing `Body` objects along as shown in the examples. Except that in my case it's being passed between 2 `Request` objects. AFAIK I'm using Hyper's APIs properly and the code compiles just fine.

**Here's the impacted code (the `tunnel_req.set_body(body);` part especially):**

Direct link to the file: [tunnel.rs](https://github.com/valeriansaliou/bloom/blob/master/src/proxy/tunnel.rs#L75)

```rust
pub struct ProxyTunnel {
    core: Core,
    client: Client<HttpConnector>,
    shards: [Option<&'static Uri>; MAX_SHARDS as usize],
}

impl ProxyTunnelBuilder {
    pub fn new() -> ProxyTunnel {
        // TODO: keep a pool of connections active? (re-use existing connectors)
        let core = Core::new().unwrap();
        let handle = core.handle();
        let client = Client::configure()
            .connector(HttpConnector::new(APP_CONF.proxy.tunnel_threads, &handle))
            .build(&handle);

        // We support only 1 shard for now.
        ProxyTunnel {
            core: core,
            client: client,
            shards: [Some(&*SHARD_URI)],
        }
    }
}

impl ProxyTunnel {
    pub fn run(
        &mut self,
        method: &Method,
        uri: &Uri,
        headers: &Headers,
        body: Body,
        shard: u8,
    ) -> Result<Response, Error> {
        if shard < MAX_SHARDS {
            // Route to target shard
            match self.shards[shard as usize] {
                Some(ref shard_uri) => {
                    let tunnel_uri = format!("{}{}", shard_uri, uri.path()).parse().unwrap();

                    let mut tunnel_req = Request::new(method.to_owned(), tunnel_uri);

                    // Forward headers
                    {
                        let tunnel_headers = tunnel_req.headers_mut();

                        tunnel_headers.clone_from(headers);
                    }

                    // Forward body
                    // TODO: blocking if non-empty, eg. if PATCH, why?
                    tunnel_req.set_body(body);

                    // TODO: debug (this works when the content-length is ==)
                    // tunnel_req.set_body(Body::from(
                    //     "{\"type\":\"online\",\"time\":{\"for\":60}}"));

                    self.core.run(self.client.request(tunnel_req))
                }
                None => Err(Error::Header),
            }
        } else {
            // Shard out of bounds
            Err(Error::Header)
        }
    }
}
```

**Below are logs I get from Hyper, running a PATCH request against a wrapped `/v1/user/availability` REST API route (notice `"Content-Length": "35"`, which is valid so I don't get why it hangs waiting forever):**

Notice the "write in body not done" part.

> (DEBUG) - Http::connect("http://127.0.0.1:3040/v1/user/availability")
> (DEBUG) - loop poll - Duration { secs: 0, nanos: 9411 }
> (DEBUG) - loop time - Instant { t: 50170802422099 }
> (DEBUG) - consuming notification queue
> (DEBUG) - resolve host="127.0.0.1", port=3040
> (DEBUG) - loop process - 1 events, Duration { secs: 0, nanos: 15928 }
> (DEBUG) - loop poll - Duration { secs: 0, nanos: 5898 }
> (DEBUG) - loop time - Instant { t: 50170802468023 }
> (DEBUG) - loop process - 1 events, Duration { secs: 0, nanos: 5191 }
> (DEBUG) - connecting to V4(127.0.0.1:3040)
> (DEBUG) - adding a new I/O source
> (DEBUG) - consuming notification queue
> (DEBUG) - scheduling direction for: 0
> (DEBUG) - blocking
> (DEBUG) - loop poll - Duration { secs: 0, nanos: 18560 }
> (DEBUG) - loop time - Instant { t: 50170802608154 }
> (DEBUG) - notifying a task handle
> (DEBUG) - loop process - 1 events, Duration { secs: 0, nanos: 13829 }
> (DEBUG) - loop poll - Duration { secs: 0, nanos: 25671 }
> (DEBUG) - loop time - Instant { t: 50170802649938 }
> (DEBUG) - loop process - 1 events, Duration { secs: 0, nanos: 8862 }
> (DEBUG) - loop poll - Duration { secs: 0, nanos: 7145 }
> (DEBUG) - loop time - Instant { t: 50170802724553 }
> (DEBUG) - consuming notification queue
> (DEBUG) - scheduling direction for: 0
> (DEBUG) - blocking
> (DEBUG) - write in body done
> (DEBUG) - encode headers = {"Host": "localhost:8080", "Bloom-Request-Shard": "0", "Authorization": "Basic NTVjZWYwZjMtNTQ2Yi00NWFlLWI3ODEtNmFkYzc4MGQ0ODM5OjIzYzM0ZjVjZDA5MTljNjg2YWE0MDI0NDBiMjI5MjQyNTY4Yjc2ZWI1MTJkMWUwM2RlNDFjNTQ5OTc5ZTRkNTY=", "If-None-Match": ""b758d6ae44a1fd57"", "Content-Type": "application/json; charset=utf-8", "Connection": "close", "User-Agent": "Paw/3.1.3 (Macintosh; OS X/10.12.6) GCDHTTPRequest", "Content-Length": "35"}
> (DEBUG) - not ready
> (DEBUG) - write in body not done
> (DEBUG) - loop process - 1 events, Duration { secs: 0, nanos: 181508 }
> (DEBUG) - loop poll - Duration { secs: 0, nanos: 7024 }
> (DEBUG) - loop time - Instant { t: 50170802916684 }
> (DEBUG) - loop process - 0 events, Duration { secs: 0, nanos: 7000 }
> (DEBUG) - loop poll - Duration { secs: 0, nanos: 7447 }
> (DEBUG) - loop time - Instant { t: 50170802934227 }
> (DEBUG) - loop process - 1 events, Duration { secs: 0, nanos: 9118 }

**If I uncomment the DEBUG part in my code, and comment the previous body set code (to mirror the body stream I'm trying to forward in my tests):**

```rust
tunnel_req.set_body(Body::from("{\"type\":\"online\",\"time\":{\"for\":60}}"));
```

...it works like a charm (and hangs if I remove some bytes from it as it's no more matching the Content-Length = 35, but this is expected and okay).

**So, question on this issue:**

* Is this a bug? Looks like a Request body cannot be forwarded to another `Request` body. Maybe it only works this way: `Request.body` <> `Response.body` (as per your example which only show how to forward a `Body` object between `Request` and `Response`).

Of course, an obvious solution would be not to proxy POST, PATCH, PUT HTTP methods to the cache backend I'm building, but for the sake of sysadmin simplicity I'd like to be able to proxy those requests also.

**Comment by valeriansaliou** [comment]
I could replicate the exact same bug by modifying the (working) example in ./examples/server.rs to:

_Note that I used my local API server for this test._

```rust
#![deny(warnings)]
extern crate futures;
extern crate hyper;
extern crate tokio_core;
extern crate pretty_env_logger;

use futures::future::FutureResult;

use hyper::Client;
use hyper::StatusCode;
use hyper::Method::Patch;
use hyper::header::ContentLength;
use hyper::server::{Http, Service, Request, Response};

struct Echo;

impl Service for Echo {
    type Request = Request;
    type Response = Response;
    type Error = hyper::Error;
    type Future = FutureResult<Response, hyper::Error>;

    fn call(&self, req: Request) -> Self::Future {
        futures::future::ok(match (req.method(), req.path()) {
            (&Patch, "/echo") => {
                let mut core = tokio_core::reactor::Core::new().unwrap();
                let handle = core.handle();
                let client = Client::new(&handle);

                let mut tunnel_req = Request::new(Patch, "http://api.crisp.chat.dev/v1/user/availability".parse().unwrap());

                if let Some(len) = req.headers().get::<ContentLength>() {
                    tunnel_req.headers_mut().set(len.clone());
                }

                tunnel_req.set_body(req.body());

                match core.run(client.request(tunnel_req)) {
                    Ok(response) => response,
                    Err(err) => panic!("could not stream: {}", err),
                }
            },
            _ => {
                Response::new().with_status(StatusCode::NotFound)
            }
        })
    }

}


fn main() {
    pretty_env_logger::init().unwrap();
    let addr = "127.0.0.1:1337".parse().unwrap();

    let server = Http::new().bind(&addr, || Ok(Echo)).unwrap();
    println!("Listening on http://{} with 1 thread.", server.local_addr().unwrap());
    server.run().unwrap();
}
```

Using cURL command:

```bash
curl -X "PATCH" "http://127.0.0.1:1337/echo" \
     -H "Content-Type: application/json; charset=utf-8" \
     -u 55cef0f3-546b-45ae-b781-6adc780d4839:23c34f5cd0919c686aa402440b229242568b76eb512d1e03de41c549979e4d56 \
     -d $'{
  "type": "online",
  "time": {
    "for": 60
  }
}'
```

**Comment by valeriansaliou** [comment]
Wouldn't that come from `futures::sync::mpsc::Receiver` <> `futures::sync::mpsc::Sender` where I'm trying to bridge `futures::sync::mpsc::Receiver` <> `futures::sync::mpsc::Receiver`?

**Comment by seanmonstar** [comment]
Note that if you're using `core.run` within the `Service::call`, that will block the thread while that specific `core` runs. This could mean the `core` driving the server will not be able to run and see events (and thus read the request body from the socket).

Instead, you should change the `Service::Future` to a different future type, instead of `FutureResult`, since sending a client request will not happen immediately.

(PS, you pasted the `Authorization` header, if those values are real, you should reset them. :))

**Comment by valeriansaliou** [comment]
Thanks @seanmonstar, will try this.

PS: on the `Authorization` header thing, woops but no worries, it's a local-only one 😅

**Comment by valeriansaliou** [comment]
Got it to work w/ this Gist as an example: https://gist.github.com/mockersf/6ae921598913c3b59799bf2a33546922 (referenced from #1189).

Should also fix a previous issue of mine #1260. Was all due to a misunderstanding on my end.

More examples on this would be nice, I guess it's a common problem for newcomers, as it's hard to learn coding proper Hyper + Tokio + Futures when you encounter all those at once, especially the Futures part.

Thanks for the help anyway! I consider this closed.
